File: app/src/main/java/com/sysu/edu/life/GymDetailFragment.java
package com.sysu.edu.life;

import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.alibaba.fastjson2.JSONArray;
import com.alibaba.fastjson2.JSONObject;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.sysu.edu.R;
import com.sysu.edu.api.AuthorizationJar;
import com.sysu.edu.api.HttpManager;
import com.sysu.edu.api.Params;
import com.sysu.edu.api.TargetUrl;
import com.sysu.edu.databinding.DialogGymReservationBinding;
import com.sysu.edu.databinding.FragmentGymDetailBinding;
import com.sysu.edu.databinding.ItemDateBinding;
import com.sysu.edu.databinding.ItemFieldDetailBinding;
import com.sysu.edu.view.RecyclerAdapter;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Locale;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;
import java.util.regex.Pattern;
import java.util.stream.IntStream;


public class GymDetailFragment extends Fragment {

    final HashMap<String, JSONObject> fee = new HashMap<>();
    HttpManager http;
    GymReservationViewModel viewModel;
    String id;
    String hash;
    DateAdapter date;
    String userId;
    String type;
    FragmentGymDetailBinding binding;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentGymDetailBinding.inflate(inflater, container, false);
        DialogGymReservationBinding dialogBinding = DialogGymReservationBinding.inflate(inflater, container, false);
        binding.date.recyclerView.setLayoutManager(new LinearLayoutManager(requireContext(), LinearLayoutManager.HORIZONTAL, false));
        date = new DateAdapter();

        BottomSheetDialog dialog = new BottomSheetDialog(requireContext());
        dialog.setContentView(dialogBinding.getRoot());
        initReservationDialog(dialogBinding);
        viewModel = new ViewModelProvider(requireActivity()).get(GymReservationViewModel.class);
        date.setAction(viewModel.position::setValue);
        Params params = new Params(this);
        binding.date.recyclerView.setAdapter(date);
        binding.date.recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (!recyclerView.canScrollHorizontally(1) && dx > 0) date.offset(7);
            }
        });
        GridLayoutManager gridLayoutManager = new GridLayoutManager(requireContext(), 4, GridLayoutManager.HORIZONTAL, false);
        binding.field.recyclerView.setLayoutManager(gridLayoutManager);
        FieldAdapter field = new FieldAdapter();
        binding.field.recyclerView.setAdapter(field);
        id = requireArguments().getString("id");
        field.setAction(_ -> {
            viewModel.selected.setValue(field.getSelected());
//                updateReservationDialog(dialogBinding, p, studentFee);
//            dialog.show();
        });
        if (viewModel.position.getValue() == null) {
            viewModel.position.postValue(0);
        }
        viewModel.position.observe(getViewLifecycleOwner(), p -> {
            if (p != null) getInfo();
        });
        http = new HttpManager(new Handler(Looper.getMainLooper()) {
            @Override
            public void handleMessage(@NonNull Message msg) {
                super.handleMessage(msg);
                if (msg.what == -1) params.toast(R.string.no_net_connected);
                else {
                    String response = (String) msg.obj;
                    if (msg.getData().getBoolean("isJSON")) {
                        switch (msg.what) {
                            case 0:
                                reset(field);
                                hash = md5(response);
                                AtomicInteger availableCapacity = new AtomicInteger();
                                AtomicInteger rows = new AtomicInteger(-1);
                                MutableLiveData<Boolean> name = new MutableLiveData<>(false);
                                JSONArray.parse(response).forEach(e -> {
                                    JSONObject item = (JSONObject) e;
                                    JSONArray timeslots = item.getJSONArray("Timeslots");
                                    if (timeslots != null) {
                                        if (rows.get() == -1) {
                                            field.add(JSONObject.of("Name", getString(R.string.time), "Type", 2));
                                            timeslots.forEach(o -> field.add(JSONObject.of("Name", String.format("%s\n%s", ((JSONObject) o).getString("Start"), ((JSONObject) o).getString("End")), "Type", 2)));
                                            name.setValue(true);
                                            rows.set(timeslots.size() + 1);
                                            gridLayoutManager.setSpanCount(rows.get());
                                        }// 第一列
                                        String fieldName = Pattern.compile("(.+)-").matcher(item.getString("VenueName")).replaceAll(""); // 第一行
                                        field.add(new JSONObject().fluentPut("VenueName", fieldName).fluentPut("Type", 0));
                                        timeslots.forEach(o -> {
                                            JSONObject jsonObject = (JSONObject) o;
                                            field.add(jsonObject.clone().fluentPut("VenueBooking", item.clone().fluentPut("Timeslots", JSONArray.of(jsonObject))).fluentPut("Type", 1).fluentPut("Venue", fieldName).fluentPut("Duration", String.format(Locale.getDefault(), "%s~%s", jsonObject.getString("Start"), jsonObject.getString("End"))));
                                            int capacity = jsonObject.getInteger("AvailableCapacity");
                                            if (capacity > 0)
                                                availableCapacity.addAndGet(capacity);
                                        });
                                        if (field.getItemCount() % rows.get() != 0)
                                            IntStream.range(0, rows.get() - field.getItemCount() % rows.get()).forEach(_ -> field.add(JSONObject.of("Type", 3)));

                                    }
                                });
                                if (viewModel.position.getValue() != null)
                                    date.setAvailableCapacity(viewModel.position.getValue(), availableCapacity.get());
                                getFee(id);
                                break;
                            case 1:
                                fee.clear();
                                JSONArray feeTemplates = JSONArray.parse(response);
                                if (feeTemplates != null)
                                    feeTemplates.forEach(e -> fee.put(((JSONObject) e).getString("UserRole"), (JSONObject) e));
                                getMe();
                                break;
                            case 2:
                                JSONArray mes = JSONArray.parseArray(response);
                                if (!mes.isEmpty()) {
                                    JSONObject me = mes.getJSONObject(0);
                                    userId = me.getString("UserId");
                                }
                                getType(id);
                                break;
                            case 3:
                                JSONArray types = JSONArray.parseArray(response);
                                if (!types.isEmpty()) {
                                    type = types.getJSONObject(0).getString("TypeIdentity");
                                }
                                break;
                            case 4:
                                JSONObject result = JSONObject.parse(response);
                                if (result.getInteger("Code") == 200) {
                                    params.toast(R.string.reserve_success);
                                } else {
                                    params.toast(result.getString("Result"));
                                }
                                // 订单编号
                                /*{"Code":200,"Result":{"Identity":"6ae40ca4-392b-4ebc-8ba2-8ec1126f54bc","Name":"tangxb6","BookingId":"#RB-44U4HTRRQ4","UserId":"tangxb6","HostKey":"24308152","VenueTypeId":"802fbfda-f9f6-41c8-9c72-685247f07c73","VenueBookings":[{"VenueId":"d2c4c59a-1d00-4c44-9a6a-24f26390227e","VenueName":"东校园游泳池","TimeSlots":[{"Date":"2026-03-18T00:00:00","Start":"19:30","End":"21:00"}]}],"Participants":[],"Status":"Accepted","Description":"东校园游泳池","Charge":-5,"IsCash":false,"CreatedAt":"2026-03-17T21:30:19.8154563+08:00","UpdatedAt":"2026-03-17T13:30:19.611245Z","ActionedBy":"tangxb6","Token":null}}*/
                                break;
                        }
                    } else if (!viewModel.authorizationManager.isAuthorized(response)) {
                        params.toast(R.string.login_warning);
                        params.gotoLogin( viewModel.authorizationManager.isAccessible() ? TargetUrl.GYM : TargetUrl.GYM_WEBVPN);
                    } else if (Pattern.compile("人机识别检测").matcher(response).find()) {
                        params.gotoLogin(viewModel.authorizationManager.isAccessible() ? TargetUrl.GYM : TargetUrl.GYM_WEBVPN);
                    } else if (!viewModel.authorizationManager.isAccessible(response)) {
                        params.toast(R.string.educational_wifi_warning);
                        getInfo();
                    }

                }
            }
        });
        http.setParams(params);
        http.setUA(viewModel.ua);
        http.setHeader(Map.of("Accept", "application/json, text/plain, */*"));
        http.setAuthorizationRequired(true);
        http.setAuthorizationJar(new AuthorizationJar(requireContext()));
        date.select(viewModel.position.getValue() == null ? 0 : viewModel.position.getValue());
        viewModel.selected.observe(getViewLifecycleOwner(), selected -> {
            field.setSelected(selected);
            JSONObject studentFee = fee.get("学生");
            if (studentFee != null) {
                binding.submit.setEnabled(!field.getSelected().isEmpty());
                if (field.getSelected().isEmpty()) {
                    binding.info.setText(getString(R.string.unselected));
                } else {
                    StringBuilder info = new StringBuilder();
                    JSONArray items = new JSONArray();
                    field.getSelected().forEach(e -> {
                        info.append(field.get(e).getString("Venue")).append(" ").append(field.get(e).getString("Duration")).append("+");
                        items.add(field.get(e).getJSONObject("VenueBooking"));
                    });
                    String venueName = field.get(new ArrayList<>(field.getSelected()).get(0)).getString("VenueName");
                    int creditFee = studentFee.getInteger("CreditFee") * field.getSelected().size();
                    binding.submit.setOnClickListener(_ -> reserve(items, venueName, creditFee));
                    binding.info.setText(String.format(Locale.getDefault(), "%s=%d元", info.deleteCharAt(info.length() - 1), creditFee));
                }
            }
        });
        return binding.getRoot();
    }

    void reset(FieldAdapter field) {
        field.clear();
        field.clearSelected();
        viewModel.selected.setValue(new HashSet<>());
    }

    void getInfo() {
        Integer positionValue = viewModel.position.getValue();
        if (positionValue != null)
            getInfo(id, date.getFormattedDate(positionValue), date.getFormattedDate(positionValue));
    }

    void getInfo(String id, String from, String to) {
        http.getRequest(viewModel.authorizationManager.getBaseUrl() + String.format("api/venue/available-slots/range?venueTypeId=%s&start=%s&end=%s", id, from, to), 0);
    }

    void getFee(String id) {
        http.getRequest(viewModel.authorizationManager.getBaseUrl() + String.format("api/venuetype/%s/feetemplates", id), 1);
    }

    void getMe() {
        http.getRequest(viewModel.authorizationManager.getBaseUrl() + "api/swimmer/me", 2);
    }

    void getType(String id) {
        http.getRequest(viewModel.authorizationManager.getBaseUrl() + "api/venue/type/" + id, 3);
    }

    public void reserve(String payload) {
        http.postRequest(viewModel.authorizationManager.getBaseUrl() + "api/BookingRequestVenue", payload, 4);
    }

    /**
     * 生成 UUID
     *
     * @return 生成的 UUID
     *
     */
    String generateUUID() {
        return UUID.randomUUID().toString();
    }

    /**
     * 生成 Token
     *
     * @param hash 哈希值
     * @return 生成的 Token
     *
     */
    String genToken(String uuid, String hash) {
        long timestamp = System.currentTimeMillis() / 1000L;
        return md5("SYSUBOOKING-" + uuid + timestamp) + "." + timestamp + "." + hash;
    }

    /**
     * 计算 MD5 哈希值
     *
     * @param input 输入字符串
     * @return 计算得到的 MD5 哈希值（十六进制小写字符串）
     *
     */
    String md5(String input) {
        try {
            StringBuilder hexString = new StringBuilder();
            for (byte b : MessageDigest.getInstance("MD5").digest(input.getBytes(StandardCharsets.UTF_8))) {
                String hex = Integer.toHexString(0xff & b);
                hexString.append(hex.length() == 1 ? "0" : hex);
            }
            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("MD5 algorithm not available", e);
        }
    }

    void initReservationDialog(DialogGymReservationBinding binding) {
        binding.field.key.setText(R.string.field);
        binding.date.key.setText(R.string.date);
        binding.time.key.setText(R.string.time);
        binding.fee.key.setText(R.string.fee);
        binding.type.key.setText(R.string.type);
    }

//    void updateReservationDialog(DialogGymReservationBinding binding, JSONObject item, JSONObject studentFee) {
//        binding.field.value.setText(item.getString("Venue"));
//        binding.date.value.setText(item.getString("Date"));
//        binding.time.value.setText(item.getString("Duration"));
//        Integer creditFee = studentFee.getInteger("CreditFee");
//        Integer cashFee = studentFee.getInteger("CashFee");
//        binding.fee.value.setText(String.format(Locale.getDefault(), "运动时￥%d或现金￥%d", creditFee, cashFee));
//        binding.type.value.setText(item.getString("Type"));
//        binding.reserve.setOnClickListener(_ -> reserve(item.getJSONArray("VenueBooking"), item.getString("VenueName"), creditFee));
//    }

    private void reserve(JSONArray items, String venueName, Integer creditFee) {
        String uuid = generateUUID();
        String time = LocalDateTime.now().atZone(ZoneId.systemDefault()).withZoneSameInstant(ZoneOffset.UTC).toString();
//            {"Identity":"561e14e8-1f6d-44d4-89e9-e9cbe0b5bb81","BookingId":"e59ec78d1b4b06f72f6860d10e0fe7d2.1773708652.b268e04b7b07d45673fdf4ad2f14c914","VenueTypeId":"802fbfda-f9f6-41c8-9c72-685247f07c73","VenueBookings":[{"VenueId":"d2c4c59a-1d00-4c44-9a6a-24f26390227e","VenueName":"东校园游泳池","TimeSlots":[{"Date":"2026-03-17T00:00:00.000Z","Start":"16:30","End":"18:00"}]}],"Participants":[],"Status":"Accepted","Description":"东校园游泳池","CreatedAt":"2026-03-17T00:50:52.552Z","UpdatedAt":"2026-03-17T00:50:52.552Z","ActionedBy":"tangxb6","IsCash":false,"Charge":5}
        JSONObject payload = JSONObject.of(
                "Identity", uuid,
                "BookingId", genToken(uuid, hash),
                "VenueTypeId", type,
                "VenueBookings", items,
                "Participants", JSONArray.of(),
                "Status", "Accepted",
                "Description", venueName,
                "CreatedAt", time,
                "UpdatedAt", time,
                "ActionedBy", userId,/*NetID*/
                "IsCash", false,
                "Charge", creditFee
        );
//        System.out.println(payload);
        reserve(payload.toJSONString());
    }


    static class DateAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

        final HashMap<Integer, Integer> availableCapacity = new HashMap<>();
        Consumer<? super Integer> action;
        int page = 7;
        int selected = -1;

        @NonNull
        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            return new RecyclerView.ViewHolder(ItemDateBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false).getRoot()) {
            };
        }

        void select(int position) {
            int tmp = selected;
            selected = position;
            action.accept(position);
            notifyItemChanged(position);
            notifyItemChanged(tmp);
        }

        public void setAction(Consumer<? super Integer> action) {
            this.action = action;
        }

        @Override
        public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
            ItemDateBinding binding = ItemDateBinding.bind(holder.itemView);
            Context context = holder.itemView.getContext();
            binding.date.setText(getDate(position));
            binding.week.setText(getWeek(context, position));
            binding.getRoot().setOnClickListener(_ -> select(position));
            Integer capacity = availableCapacity.getOrDefault(position, -1);
            binding.getRoot().setChecked(position == selected);
            binding.availableCapacity.setText(capacity != null && capacity >= 0 ? String.format(Locale.getDefault(), "%d", capacity) : "");
        }

        public void setAvailableCapacity(int position, int i) {
            availableCapacity.put(position, i);
            notifyItemChanged(position);
        }

        @Override
        public int getItemCount() {
            return page;
        }

        public void offset(int offset) {
            page += offset;
            notifyItemRangeInserted(page - 1, offset);
        }

        private String getDate(int distanceDay, String pattern) {
            return LocalDate.now().plusDays(distanceDay).format(DateTimeFormatter.ofPattern(pattern));
        }

        public String getDate(int distanceDay) {
            return getDate(distanceDay, "M月dd日");
        }

        public String getFormattedDate(int distanceDay) {
            return getDate(distanceDay, "M-dd");
        }

        String getWeek(Context context, int distanceDay) {
            int week = LocalDate.now().plusDays(distanceDay).getDayOfWeek().getValue();
            return context.getResources().getStringArray(R.array.weeks)[week - 1];
        }

    }

    static class FieldAdapter extends RecyclerAdapter<JSONObject> {

        Consumer<? super Integer> action;
        HashSet<Integer> selected = new HashSet<>();

        @NonNull
        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            return new RecyclerView.ViewHolder(ItemFieldDetailBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false).getRoot()) {
            };
        }

        void setAction(Consumer<? super Integer> action) {
            this.action = action;
        }

        void clearSelected() {
            HashSet<Integer> s = new HashSet<>(selected);
            selected.clear();
            s.forEach(this::notifyItemChanged);
        }

        HashSet<Integer> getSelected() {
            return selected;
        }

        public void setSelected(HashSet<Integer> selected) {
            if (!this.selected.equals(selected)) {
                this.selected = selected;
                selected.forEach(this::notifyItemChanged);
            }
        }

        @Override
        public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int pos) {
            int position = holder.getBindingAdapterPosition();
            ItemFieldDetailBinding binding = ItemFieldDetailBinding.bind(holder.itemView);
            binding.fieldDetail.setAlpha(1.0f);
            JSONObject item = get(position);
            Context context = holder.itemView.getContext();
            binding.getRoot().setOnClickListener(_ -> {
                if (item.getInteger("Type") == 1 && item.getInteger("AvailableCapacity") > 0) {
                    if (selected.contains(position)) {
                        selected.remove(position);
                    } else {
                        selected.add(position);
                    }
                    action.accept(position);
                    notifyItemChanged(position);
                }
            });
            switch (item.getInteger("Type")) {
                case 0 ->
                        binding.fieldDetail.setText(String.format(Locale.getDefault(), "%s", item.getString("VenueName")));
                case 2 ->
                        binding.fieldDetail.setText(String.format(Locale.getDefault(), "%s", item.getString("Name")));
                case 1 -> {
                    if (item.getInteger("AvailableCapacity") > 0) {
                        binding.fieldDetail.setText(context.getString(R.string.reservable));
                    } else {
                        binding.fieldDetail.setText(context.getString(R.string.reserved));
                        binding.fieldDetail.setAlpha(0.5f);
                    }
                    binding.getRoot().setChecked(selected.contains(position));
                }
                default -> binding.fieldDetail.setText("");
            }
        }
    }
}

File: app/src/main/java/com/sysu/edu/life/GymDetailFragment.kt
package com.sysu.edu.life

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.alibaba.fastjson2.JSONArray
import com.alibaba.fastjson2.JSONObject
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.sysu.edu.BaseFragment
import com.sysu.edu.R
import com.sysu.edu.databinding.DialogGymReservationBinding
import com.sysu.edu.databinding.FragmentGymDetailBinding
import com.sysu.edu.databinding.ItemDateBinding
import com.sysu.edu.databinding.ItemFieldDetailBinding
import com.sysu.edu.model.GymModel
import com.sysu.edu.view.RecyclerAdapter
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.Locale
import java.util.UUID
import java.util.function.Consumer
import java.util.regex.Pattern

class GymDetailFragment : BaseFragment() {
	val fee: MutableMap<String?, JSONObject?> = mutableMapOf()
	var viewModel: GymReservationViewModel? = null
	var id: String? = null
	var hash: String? = null
	var dateAdapter: DateAdapter? = null
	var userId: String? = null
	var type: String? = null
	val model: GymModel by lazy {
		GymModel(requireContext())
	}
	
	override fun onCreateView(inflater: LayoutInflater,
	                          container: ViewGroup?,
	                          savedInstanceState: Bundle?): View {
		super.onCreateView(inflater, container, savedInstanceState)
		viewModel = ViewModelProvider(requireActivity())[GymReservationViewModel::class.java]
		id = requireArguments().getString("id")
		val gridLayoutManager = GridLayoutManager(requireContext(), 4, GridLayoutManager.HORIZONTAL, false)
		val fieldAdapter = FieldAdapter().apply {
			action = {
				viewModel!!.selected.value = selected
			}
		}
		dateAdapter = DateAdapter().apply {
			action = { value: Int? -> viewModel!!.position.value = value }
			select((viewModel!!.position.value ?: 0))
		}
		val binding = FragmentGymDetailBinding.inflate(inflater, container, false).apply {
			date.recyclerView.adapter = dateAdapter
			date.recyclerView.layoutManager = LinearLayoutManager(requireContext(), LinearLayoutManager.HORIZONTAL, false)
			date.recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
				override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
					super.onScrolled(recyclerView, dx, dy)
					if (!recyclerView.canScrollHorizontally(1) && dx > 0) dateAdapter?.offset(7)
				}
			})
			field.recyclerView.layoutManager = gridLayoutManager
			field.recyclerView.adapter = fieldAdapter
		}
		val dialogBinding = DialogGymReservationBinding.inflate(inflater, container, false).apply {
			field.key.setText(R.string.field)
			date.key.setText(R.string.date)
			time.key.setText(R.string.time)
			fee.key.setText(R.string.fee)
			type.key.setText(R.string.type)
		}
		val dialog = BottomSheetDialog(requireContext())
		dialog.setContentView(dialogBinding.root)
		if (viewModel!!.position.value == null) viewModel!!.position.postValue(0)
		viewModel!!.position.observe(viewLifecycleOwner) { p: Int? ->
			if (p != null) info
		}
		model.message.observe(viewLifecycleOwner) { (code, response) ->
			println(response)
			when (code) {
				0 -> {
					reset(fieldAdapter)
					hash = md5("$response")
					var availableCapacity = 0
					var rows = -1
					val name = MutableLiveData(false)
					response.getJSONArray("data").forEach(Consumer { item: Any? ->
						val timeslots = (item as JSONObject).getJSONArray("Timeslots")
						if (timeslots != null) {
							if (rows == -1) {
								fieldAdapter.add(JSONObject.of("Name", getString(R.string.time), "Type", 2))
								timeslots.forEach { o: Any? -> fieldAdapter.add(JSONObject.of("Name", "${(o as JSONObject).getString("Start")}\n${o.getString("End")}", "Type", 2)) }
								name.value = true
								rows = timeslots.size + 1
								gridLayoutManager.spanCount = rows
							} // 第一列
							val fieldName = Pattern.compile("(.+)-")
								.matcher(item.getString("VenueName"))
								.replaceAll("") // 第一行
							fieldAdapter.add(JSONObject().fluentPut("VenueName", fieldName)
												 .fluentPut("Type", 0))
							timeslots.forEach { data: Any? ->
								fieldAdapter.add((data as JSONObject).clone()
													 .fluentPut("VenueBooking", item.clone()
														 .fluentPut("Timeslots", JSONArray.of(data)))
													 .fluentPut("Type", 1)
													 .fluentPut("Venue", fieldName)
													 .fluentPut("Duration", "${data.getString("Start")}~${data.getString("End")}"))
								data.getInteger("AvailableCapacity")?.let {
									availableCapacity += it
								}
							}
							if (fieldAdapter.itemCount % rows != 0) (0..<(rows - fieldAdapter.itemCount % rows)).forEach { _ -> fieldAdapter.add(JSONObject.of("Type", 3)) }
						}
					})
					if (viewModel!!.position.value != null) dateAdapter!!.setAvailableCapacity(viewModel!!.position.value!!, availableCapacity)
					getFee(id!!)
				}
				1 -> {
					fee.clear()
					response.getJSONArray("data")?.run {
						forEach { fee[(it as JSONObject).getString("UserRole")] = it }
					}
					me
				}
				2 -> {
					response.getJSONArray("data")?.takeUnless { it.isEmpty() }?.let {
						userId = it.getJSONObject(0).getString("UserId")
					}
					getType(id)
				}
				3 -> {
					response.getJSONArray("data")?.takeUnless { it.isEmpty() }?.let {
						type = it.getJSONObject(0).getString("TypeIdentity")
					}
				}
				4 -> {
					with(response.getJSONObject("data")) {
						if (getInteger("Code") == 200) config.toast(R.string.reserve_success)
						else config.toast(getString("Result")) // 订单编号
					}
				}
			}
		}
		viewModel!!.selected.observe(viewLifecycleOwner, Observer { selected: MutableSet<Int>? ->
			fieldAdapter.selected = selected
			val studentFee = fee["学生"]
			if (studentFee != null) {
				fieldAdapter.selected?.isEmpty().let {
					binding.submit.setEnabled(it == false)
					if (it == true) binding.info.text = getString(R.string.unselected)
					else {
						val info = StringBuilder()
						val items = JSONArray()
						fieldAdapter.selected?.forEach { e: Int? ->
							info.append(fieldAdapter.get(e!!).getString("Venue"))
								.append(" ")
								.append(fieldAdapter.get(e).getString("Duration"))
								.append("+")
							items.add(fieldAdapter.get(e).getJSONObject("VenueBooking"))
						}
						val venueName = fieldAdapter.get(fieldAdapter.selected?.toList()?.get(0)!!)
							.getString("VenueName")
						val creditFee = studentFee.getInteger("CreditFee") * fieldAdapter.selected?.size!!
						binding.submit.setOnClickListener { reserve(items, venueName, creditFee) }
						binding.info.text = String.format(Locale.getDefault(), "%s=%d元", info.deleteCharAt(info.length - 1), creditFee)
					}
				}
			}
		})
		return binding.root
	}
	
	fun reset(field: FieldAdapter) {
		field.clear()
		field.clearSelected()
		viewModel!!.selected.value = HashSet()
	}
	
	val info: Unit
		get() {
			viewModel!!.position.value?.let {
				getInfo(id!!, dateAdapter!!.getFormattedDate(it), dateAdapter!!.getFormattedDate(it))
			}
		}
	
	fun getInfo(id: String, from: String?, to: String?) {
		model.addAndNext("api/venue/available-slots/range?venueTypeId=$id&start=$from&end=$to", 0)
	}
	
	fun getFee(id: String) {
		model.addAndNext("api/venuetype/$id/feetemplates", 1)
	}
	
	val me: Unit
		get() {
			model.addAndNext("api/swimmer/me", 2)
		}
	
	fun getType(id: String?) {
		model.addAndNext("api/venue/type/$id", 3)
	}
	
	fun reserve(payload: String?) {
		model.addAndNext("api/BookingRequestVenue", payload, 4)
	}
	
	/**
	 * 生成 UUID
	 *
	 * @return 生成的 UUID
	 */
	fun generateUUID(): String {
		return UUID.randomUUID().toString()
	}
	
	/**
	 * 生成 Token
	 *
	 * @param hash 哈希值
	 * @return 生成的 Token
	 */
	fun genToken(uuid: String?, hash: String?): String {
		val timestamp = System.currentTimeMillis() / 1000L
		return md5("SYSUBOOKING-$uuid$timestamp") + "." + timestamp + "." + hash
	}
	
	/**
	 * 计算 MD5 哈希值
	 *
	 * @param input 输入字符串
	 * @return 计算得到的 MD5 哈希值（十六进制小写字符串）
	 */
	fun md5(input: String): String {
		try {
			val hexString = StringBuilder()
			(MessageDigest.getInstance("MD5")
				.digest(input.toByteArray(StandardCharsets.UTF_8))).forEach {
					hexString.append(Integer.toHexString(0xff and it.toInt())
										 .takeUnless { it1 -> it1.length == 1 } ?: "0")
				}
			return "$hexString"
		} catch (e: NoSuchAlgorithmException) {
			throw RuntimeException("MD5 algorithm not available", e)
		}
	}
	
	//    void updateReservationDialog(DialogGymReservationBinding binding, JSONObject item, JSONObject studentFee) {
	//        binding.field.value.setText(item.getString("Venue"));
	//        binding.date.value.setText(item.getString("Date"));
	//        binding.time.value.setText(item.getString("Duration"));
	//        Integer creditFee = studentFee.getInteger("CreditFee");
	//        Integer cashFee = studentFee.getInteger("CashFee");
	//        binding.fee.value.setText(String.format(Locale.getDefault(), "运动时￥%d或现金￥%d", creditFee, cashFee));
	//        binding.type.value.setText(item.getString("Type"));
	//        binding.reserve.setOnClickListener(_ -> reserve(item.getJSONArray("VenueBooking"), item.getString("VenueName"), creditFee));
	//    }
	private fun reserve(items: JSONArray?, venueName: String?, creditFee: Int?) {
		val uuid = generateUUID()
		val time = LocalDateTime.now()
			.atZone(ZoneId.systemDefault())
			.withZoneSameInstant(ZoneOffset.UTC)
			.toString()
		val payload = JSONObject.of("Identity", uuid, "BookingId", genToken(uuid, hash), "VenueTypeId", type, "VenueBookings", items, "Participants", JSONArray.of(), "Status", "Accepted", "Description", venueName, "CreatedAt", time, "UpdatedAt", time, "ActionedBy", userId,  /*NetID*/
		                            "IsCash", false, "Charge", creditFee) //        System.out.println(payload);
		reserve(payload.toJSONString())
	}
	
	class DateAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder?>() {
		val availableCapacity: MutableMap<Int?, Int?> = mutableMapOf()
		var action: ((Int) -> Unit)? = null
		var page: Int = 7
		var selected: Int = -1
		override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
			return object :
				RecyclerView.ViewHolder(ItemDateBinding.inflate(LayoutInflater.from(parent.context), parent, false).root) {}
		}
		
		fun select(position: Int) {
			val tmp = selected
			selected = position
			action!!(position)
			notifyItemChanged(position)
			notifyItemChanged(tmp)
		}
		
		override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
			val binding = ItemDateBinding.bind(holder.itemView).apply {
				date.text = getDate(position)
				week.text = getWeek(holder.itemView.context, position)
				root.setOnClickListener { select(position) }
				root.isChecked = position == selected
			}
			binding.availableCapacity.text = availableCapacity.getOrDefault(position, -1)
				?.let { if (it >= 0) "$it" else "" }
		}
		
		fun setAvailableCapacity(position: Int, i: Int) {
			availableCapacity[position] = i
			notifyItemChanged(position)
		}
		
		override fun getItemCount(): Int = page
		fun offset(offset: Int) {
			page += offset
			notifyItemRangeInserted(page - 1, offset)
		}
		
		private fun getDate(distanceDay: Int, pattern: String?): String? {
			return LocalDate.now()
				.plusDays(distanceDay.toLong())
				.format(DateTimeFormatter.ofPattern(pattern))
		}
		
		fun getDate(distanceDay: Int): String? = getDate(distanceDay, "M月dd日")
		fun getFormattedDate(distanceDay: Int): String? = getDate(distanceDay, "M-dd")
		fun getWeek(context: Context, distanceDay: Int): String? =
			context.resources.getStringArray(R.array.weeks)[LocalDate.now()
				.plusDays(distanceDay.toLong())
				.getDayOfWeek().value - 1]
	}
	
	class FieldAdapter : RecyclerAdapter<JSONObject>() {
		var action: ((Int) -> Unit)? = null
		var selected: MutableSet<Int>? = null
			set(selected) {
				if (field != selected) {
					field = selected?.also {
						it.forEach { i -> notifyItemChanged(i) }
					}
				}
			}
		
		override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
			return object :
				RecyclerView.ViewHolder(ItemFieldDetailBinding.inflate(LayoutInflater.from(parent.context), parent, false).root) {}
		}
		
		fun clearSelected() {
			val s = selected
			selected?.clear()
			s?.forEach { notifyItemChanged(it) }
		}
		
		override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
			val pos = holder.getBindingAdapterPosition()
			val item = get(pos)
			ItemFieldDetailBinding.bind(holder.itemView).apply {
				val context = root.context
				fieldDetail.alpha = 1.0f
				root.setOnClickListener {
					if (item.getInteger("Type") == 1 && item.getInteger("AvailableCapacity") > 0) {
						if (selected!!.contains(pos)) selected!!.remove(pos)
						else selected!!.add(pos)
						action!!(pos)
						notifyItemChanged(pos)
					}
				}
				when (item.getInteger("Type")) {
					0 -> fieldDetail.text = item.getString("VenueName", "")
					2 -> fieldDetail.text = item.getString("Name", "")
					1 -> {
						if (item.getInteger("AvailableCapacity") > 0) {
							fieldDetail.text = "${context.getString(R.string.reservable)}/${item.getString("AvailableCapacity", "")}"
						}
						else {
							fieldDetail.setText(R.string.reserved)
							fieldDetail.setAlpha(0.5f)
						}
						root.isChecked = selected!!.contains(pos)
					}
					else -> fieldDetail.text = ""
				}
			}
		}
	}
}